exporter.go

```diff
@@ -122,12 +123,59 @@ func (f *fluentforwardExporter) convertLogToMap(lr plog.LogRecord) map[string]in
 			}
 		}
 	}
+	if f.config.KubernetesMetadata != nil {
+		key := f.config.KubernetesMetadata.Key
+		if f.config.KubernetesMetadata.Key == "" {
+			key = "kubernetes"
+		}
+		var namespace, container, pod, node string
+		var labels map[string]string
+		res.Resource().Attributes().Range(func(k string, v pcommon.Value) bool {
+			if k == "k8s.namespace.name" {
+				namespace = v.AsString()
+				return true
+			}
+			if k == "k8s.container.name" {
+				container = v.AsString()
+			}
+			if k == "k8s.pod.name" {
+				pod = v.AsString()
+			}
+			if k == "k8s.node.name" {
+				node = v.AsString()
+			}
+			if f.config.KubernetesMetadata.IncludePodLabels && strings.HasPrefix(k, "k8s.pod.labels.") {
+				if labels == nil {
+					labels = make(map[string]string)
+				}
+				labelKey := strings.TrimPrefix(k, "k8s.pod.labels.")
+				labels[labelKey] = v.AsString()
+			}
+			return true
+		})
+
+		k8sMetadata := map[string]interface{}{
+			"namespace_name": namespace,
+			"container_name": container,
+			"pod_name":       pod,
+			"host":           node,
+		}
+
+		if f.config.KubernetesMetadata.IncludePodLabels {
+			k8sMetadata["labels"] = labels
+		}
+
+		m[key] = k8sMetadata
+	}
+
+	f.settings.Logger.Debug(fmt.Sprintf("message %+v", m))
+
```

Comment:
This code is for use only with [processor/k8sattributesprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor) 

I see 2 things:
1. Put log Resources into a nested object, like from flat to a sub json document
2. Rename fields, for example from `k8s.namespace.name` to `namespace_name` in the parent object key labeled by the `Key` property.

First idea for me is to not rename, if you want to rename, you must use the [processor/attributesprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributesprocessor) after the `k8sattributesprocessor`.

Second idea for me is to not be kubernetes specific. Whatever the [ResourceLogs](https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/plog/generated_resourcelogs.go#L22) come from.

Example, to get the value in the log line as flat (not in a nested json object), you just need to add it to the `default_labels_enabled` property, example:

```yaml
exporters:
  fluentforward:
    endpoint:
      tcp_addr: localhost:24224
    default_labels_enabled:
      k8s.namespace.name: true
```

I am wondering in which scenario on your side creating a sub json object is needed ?

Reply:
> First idea for me is to not rename, if you want to rename, you must use the [processor/attributesprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributesprocessor) after the `k8sattributesprocessor`.

That is exactly what we are doing!
You can check it out here:
- https://github.com/kube-logging/telemetry-controller/blob/main/internal/controller/telemetry/pipeline/components/processor/k8s_attributes_processor.go
- https://github.com/kube-logging/telemetry-controller/blob/e9335cec53285bbc36f19c244f8a48c7a58d4412/internal/controller/telemetry/pipeline/components/processor/attributes_processor.go#L75

Also sharing with you an example configuration visualized using [OtelBin](https://www.otelbin.io/#config=connectors%3A*N__count%2Foutput*_metrics%3A*N____logs%3A*N______telemetry*_controller*_output*_log*_count%3A*N________attributes%3A*N__________-_key%3A_tenant*N__________-_key%3A_subscription*N__________-_key%3A_exporter*N________description%3A_The_number_of_logs_sent_out_from_each_exporter.*N________resource*_attributes%3A*N__________-_key%3A_k8s.namespace.name*N__________-_key%3A_k8s.node.name*N__________-_key%3A_k8s.container.name*N__________-_key%3A_k8s.pod.name*N__________-_key%3A_k8s.pod.labels.app.kubernetes.io%2Fname*N__________-_key%3A_k8s.pod.labels.app*N__count%2Ftenant*_metrics%3A*N____logs%3A*N______telemetry*_controller*_tenant*_log*_count%3A*N________attributes%3A*N__________-_key%3A_tenant*N________description%3A_The_number_of_logs_from_each_tenant_pipeline.*N________resource*_attributes%3A*N__________-_key%3A_k8s.namespace.name*N__________-_key%3A_k8s.node.name*N__________-_key%3A_k8s.container.name*N__________-_key%3A_k8s.pod.name*N__________-_key%3A_k8s.pod.labels.app.kubernetes.io%2Fname*N__________-_key%3A_k8s.pod.labels.app*N__routing%2Fsubscription*_example-tenant-ns*_subscription-sample-1*_outputs%3A*N____table%3A*N______-_condition%3A_%22true%22*N________pipelines%3A*N__________-_logs%2Foutput*_example-tenant-ns*_subscription-sample-1*_collector*_otlp-test-output*N__routing%2Fsubscription*_example-tenant-ns*_subscription-sample-2*_outputs%3A*N____table%3A*N______-_condition%3A_%22true%22*N________pipelines%3A*N__________-_logs%2Foutput*_example-tenant-ns*_subscription-sample-2*_collector*_otlp-test-output-2*N__routing%2Ftenant*_example-tenant*_subscriptions%3A*N____table%3A*N______-_condition%3A_%22true%22*N________pipelines%3A*N__________-_logs%2Ftenant*_example-tenant*_subscription*_example-tenant-ns*_subscription-sample-1*N__________-_logs%2Ftenant*_example-tenant*_subscription*_example-tenant-ns*_subscription-sample-2*Nexporters%3A*N__debug%3A*N____verbosity%3A_detailed*N__otlp%2Fcollector*_otlp-test-output%3A*N____endpoint%3A_receiver-collector.example-tenant-ns.svc.cluster.local%3A4317*N____retry*_on*_failure%3A*N______enabled%3A_true*N______max*_elapsed*_time%3A_0*N____sending*_queue%3A*N______enabled%3A_true*N______queue*_size%3A_1000*N______storage%3A_file*_storage%2Fdefault*N____tls%3A*N______insecure%3A_true*N__otlp%2Fcollector*_otlp-test-output-2%3A*N____endpoint%3A_receiver-collector.example-tenant-ns.svc.cluster.local%3A4317*N____retry*_on*_failure%3A*N______enabled%3A_true*N______max*_elapsed*_time%3A_0*N____sending*_queue%3A*N______enabled%3A_true*N______queue*_size%3A_1000*N______storage%3A_file*_storage%2Fdefault*N____tls%3A*N______insecure%3A_true*N__prometheus%2Fmessage*_metrics*_exporter%3A*N____endpoint%3A_%3A9999*Nextensions%3A*N__file*_storage%2Fdefault%3A*N____create*_directory%3A_true*N____directory%3A_%2Fvar%2Flib%2Fotelcol%2Ffile*_storage*Nprocessors%3A*N__attributes%2Fexporter*_name*_otlp-test-output%3A*N____actions%3A*N______-_action%3A_insert*N________key%3A_exporter*N________value%3A_otlp%2Fcollector*_otlp-test-output*N__attributes%2Fexporter*_name*_otlp-test-output-2%3A*N____actions%3A*N______-_action%3A_insert*N________key%3A_exporter*N________value%3A_otlp%2Fcollector*_otlp-test-output-2*N__attributes%2Fmetricattributes%3A*N____actions%3A*N______-_action%3A_insert*N________from*_attribute%3A_k8s.pod.labels.app*N________key%3A_app*N______-_action%3A_insert*N________from*_attribute%3A_k8s.node.name*N________key%3A_host*N______-_action%3A_insert*N________from*_attribute%3A_k8s.namespace.name*N________key%3A_namespace*N______-_action%3A_insert*N________from*_attribute%3A_k8s.container.name*N________key%3A_container*N______-_action%3A_insert*N________from*_attribute%3A_k8s.pod.name*N________key%3A_pod*N__attributes%2Fsubscription*_subscription-sample-1%3A*N____actions%3A*N______-_action%3A_insert*N________key%3A_subscription*N________value%3A_subscription-sample-1*N__attributes%2Fsubscription*_subscription-sample-2%3A*N____actions%3A*N______-_action%3A_insert*N________key%3A_subscription*N________value%3A_subscription-sample-2*N__attributes%2Ftenant*_example-tenant%3A*N____actions%3A*N______-_action%3A_insert*N________key%3A_tenant*N________value%3A_example-tenant*N__deltatocumulative%3A_%7B%7D*N__k8sattributes%3A*N____auth*_type%3A_serviceAccount*N____extract%3A*N______labels%3A*N________-_from%3A_pod*N__________key*_regex%3A_.***N__________tag*_name%3A_all*_labels*N______metadata%3A*N________-_k8s.pod.name*N________-_k8s.pod.uid*N________-_k8s.deployment.name*N________-_k8s.namespace.name*N________-_k8s.node.name*N________-_k8s.pod.start*_time*N____passthrough%3A_false*N____pod*_association%3A*N______-_sources%3A*N__________-_from%3A_resource*_attribute*N____________name%3A_k8s.namespace.name*N__________-_from%3A_resource*_attribute*N____________name%3A_k8s.pod.name*N__memory*_limiter%3A*N____check*_interval%3A_1s*N____limit*_percentage%3A_75*N____spike*_limit*_mib%3A_25*Nreceivers%3A*N__filelog%2Fexample-tenant%3A*N____exclude%3A*N______-_%2Fvar%2Flog%2Fpods%2F**%2Fotc-container%2F**.log*N____include%3A*N______-_%2Fvar%2Flog%2Fpods%2Fexample-tenant-ns*_**%2F**%2F**.log*N____include*_file*_name%3A_false*N____include*_file*_path%3A_true*N____operators%3A*N______-_id%3A_get-format*N________routes%3A*N__________-_expr%3A_body_matches_%22%5E*B*B%7B%22*N____________output%3A_parser-docker*N__________-_expr%3A_body_matches_%22%5E%5B%5E_Z%5D*PZ%22*N____________output%3A_parser-containerd*N________type%3A_router*N______-_id%3A_parser-containerd*N________output%3A_extract*_metadata*_from*_filepath*N________regex%3A_%5E*C*QP*Ltime*G%5B%5E_%5EZ%5D*PZ*D_*C*QP*Lstream*Gstdout%7Cstderr*D_*C*QP*Llogtag*G%5B%5E_%5D***D_*Q*C*QP*Llog*G.***D*S*N________timestamp%3A*N__________layout%3A_*%22*.Y-*.m-*.dT*.H%3A*.M%3A*.S.*.LZ*%22*N__________parse*_from%3A_attributes.time*N________type%3A_regex*_parser*N______-_id%3A_parser-docker*N________output%3A_extract*_metadata*_from*_filepath*N________timestamp%3A*N__________layout%3A_*%22*.Y-*.m-*.dT*.H%3A*.M%3A*.S.*.LZ*%22*N__________parse*_from%3A_attributes.time*N________type%3A_json*_parser*N______-_cache%3A*N__________size%3A_128*N________id%3A_extract*_metadata*_from*_filepath*N________parse*_from%3A_attributes%5B%22log.file.path%22%5D*N________regex%3A_%5E.***B%2F*C*QP*Lnamespace*G%5B%5E*_%5D*P*D*_*C*QP*Lpod*_name*G%5B%5E*_%5D*P*D*_*C*QP*Luid*G%5Ba-f0-9-%5D*P*D*B%2F*C*QP*Lcontainer*_name*G%5B%5E*B%2F%5D*P*D*B%2F*C*QP*Lrestart*_count*G*Bd*P*D*B.log*S*N________type%3A_regex*_parser*N______-_from%3A_attributes.log*N________to%3A_body*N________type%3A_move*N______-_from%3A_attributes.stream*N________to%3A_attributes%5B%22log.iostream%22%5D*N________type%3A_move*N______-_from%3A_attributes.container*_name*N________to%3A_resource%5B%22k8s.container.name%22%5D*N________type%3A_move*N______-_from%3A_attributes.namespace*N________to%3A_resource%5B%22k8s.namespace.name%22%5D*N________type%3A_move*N______-_from%3A_attributes.pod*_name*N________to%3A_resource%5B%22k8s.pod.name%22%5D*N________type%3A_move*N______-_from%3A_attributes.restart*_count*N________to%3A_resource%5B%22k8s.container.restart*_count%22%5D*N________type%3A_move*N______-_from%3A_attributes.uid*N________to%3A_resource%5B%22k8s.pod.uid%22%5D*N________type%3A_move*N____retry*_on*_failure%3A*N______enabled%3A_true*N______max*_elapsed*_time%3A_0*N____start*_at%3A_end*N____storage%3A_file*_storage%2Fdefault*Nservice%3A*N__extensions%3A*N____-_file*_storage%2Fdefault*N__pipelines%3A*N____logs%2Foutput*_example-tenant-ns*_subscription-sample-1*_collector*_otlp-test-output%3A*N______exporters%3A*N________-_otlp%2Fcollector*_otlp-test-output*N________-_count%2Foutput*_metrics*N______processors%3A*N________-_memory*_limiter*N________-_attributes%2Fexporter*_name*_otlp-test-output*N______receivers%3A*N________-_routing%2Fsubscription*_example-tenant-ns*_subscription-sample-1*_outputs*N____logs%2Foutput*_example-tenant-ns*_subscription-sample-2*_collector*_otlp-test-output-2%3A*N______exporters%3A*N________-_otlp%2Fcollector*_otlp-test-output-2*N________-_count%2Foutput*_metrics*N______processors%3A*N________-_memory*_limiter*N________-_attributes%2Fexporter*_name*_otlp-test-output-2*N______receivers%3A*N________-_routing%2Fsubscription*_example-tenant-ns*_subscription-sample-2*_outputs*N____logs%2Ftenant*_example-tenant%3A*N______exporters%3A*N________-_routing%2Ftenant*_example-tenant*_subscriptions*N________-_count%2Ftenant*_metrics*N______processors%3A*N________-_memory*_limiter*N________-_k8sattributes*N________-_attributes%2Ftenant*_example-tenant*N______receivers%3A*N________-_filelog%2Fexample-tenant*N____logs%2Ftenant*_example-tenant*_subscription*_example-tenant-ns*_subscription-sample-1%3A*N______exporters%3A*N________-_routing%2Fsubscription*_example-tenant-ns*_subscription-sample-1*_outputs*N______processors%3A*N________-_memory*_limiter*N________-_attributes%2Fsubscription*_subscription-sample-1*N______receivers%3A*N________-_routing%2Ftenant*_example-tenant*_subscriptions*N____logs%2Ftenant*_example-tenant*_subscription*_example-tenant-ns*_subscription-sample-2%3A*N______exporters%3A*N________-_routing%2Fsubscription*_example-tenant-ns*_subscription-sample-2*_outputs*N______processors%3A*N________-_memory*_limiter*N________-_attributes%2Fsubscription*_subscription-sample-2*N______receivers%3A*N________-_routing%2Ftenant*_example-tenant*_subscriptions*N____metrics%2Foutput%3A*N______exporters%3A*N________-_prometheus%2Fmessage*_metrics*_exporter*N______processors%3A*N________-_memory*_limiter*N________-_deltatocumulative*N________-_attributes%2Fmetricattributes*N______receivers%3A*N________-_count%2Foutput*_metrics*N____metrics%2Ftenant%3A*N______exporters%3A*N________-_prometheus%2Fmessage*_metrics*_exporter*N______processors%3A*N________-_memory*_limiter*N________-_deltatocumulative*N________-_attributes%2Fmetricattributes*N______receivers%3A*N________-_count%2Ftenant*_metrics*N__telemetry%3A*N____metrics%3A*N______address%3A_0.0.0.0%3A8888*N______level%3A_detailed*N______readers%3A*N________-_pull%3A*N____________exporter%3A*N______________prometheus%3A*N________________host%3A_%22%22*N________________port%3A_8888*N%7E).

- I would also like to emphasize that the `k8sattributesprocessor` and the `attributesprocessors` in the pipelines generated by our operator are always present, and cannot be turned off. Mainly to be able to provide users with consistent metrics about their tenants.

Reply:
For the renaming topic, taking one field for example for the discussion but it applies to the four fields.

Why extracting `k8s.container.name` and putting its value at in the log field labeled: `container_name` ? 

With my short configuration example above, you can keep the attribute labeled `k8s.container.name` and its value in the log.

And you can do the same with all the others `k8s.pod.name` etc...

Reply:
The project I've sent the code-snippets from, soon will be available to use together, with one of our other operators: https://github.com/kube-logging/logging-operator

This operator historically used fluentbit for log-collection, which sent the logs in that structure. So we opted to use the same approach with otelcol aswell for compatibility reasons.
This also ensures we don't accidentally break components on other levels of the pipeline.

Reply:
I have not checked all the log processors, but I think you can get the same log output from fluentforwardexporter using processors and adding a `default_labels_enabled` like I suggested above.

It is super important to me to not bake strongly kubernetes dependent logic and attributes within the exporter. Have you seen other otelcol exporters doing that (I am not sure) ?

Reply:
I totally understand if you don't want to include this because of k8s dependent logic. It was just a proposal from our-side. Feel free to close this. :)

Reply:
We could change the exporter logic by using a denylist instead of an allowlist with the `default_labels_enabled` property. But works good like it is for now.

I am closing the PR.